File: 120 Triangle.cpp
#include <iostream>
#include <vector>
#include <algorithm>
/*
120 Triangle
https://oj.leetcode.com/problems/triangle/
*/
typedef std::vector<int> VEC;
typedef std::vector<VEC> MAT;

class Solution {
public:
    int minimumTotal(MAT& mat)
    {
        if(mat.size() < 1) return 0;
        VEC& prev = mat[0];
        
        for(size_t j = 1;j < mat.size();j++)
        {
            VEC& curr = mat[j];
            
            curr[0] += prev[0];
            for(std::size_t i = 0;i < j - 1;i++)
            {
            	curr[i+1] += std::min(prev[i],prev[i+1]);
            }
            curr[j] += prev[j-1];
            prev = curr;
        }
        return *std::min_element(prev.begin(),prev.end());
    }
};

File: 120 Triangle.py
'''
120. Triangle
https://leetcode.com/problems/triangle/

Given a triangle, find the minimum path sum from top to bottom.
Each step you may move to adjacent numbers on the row below.

For example, given the following triangle

[
     [2],
    [3,4],
   [6,5,7],
  [4,1,8,3]
]
The minimum path sum from top to bottom is 11 (i.e., 2 + 3 + 5 + 1 = 11).

Note:

Bonus point if you are able to do this using only O(n) extra space,
where n is the total number of rows in the triangle.
'''
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if not triangle:
            return 0
        
        dp = [triangle[0][0]]
        for i in range(1, len(triangle)):
            row = triangle[i]
            newDp = [0] * len(row)
            
            newDp[0] = dp[0] + row[0]
            newDp[-1] = dp[-1] + row[-1]
            for j in range(1, len(row) - 1):
                newDp[j] = min(dp[j-1], dp[j]) + row[j]
                
            dp = newDp
            
        return min(dp)
